File: Daily Challenge/February 2025/Max Sum of a Pair With Equal Sum of Digits.cpp
class Solution {
public:
  int maximumSum(vector<int>& nums) {
    unordered_map<int, int> digitSumOfValue;
    int res = -1;

    for (int x : nums) {
      int digitSum = this->digitSum(x);
      if (digitSumOfValue.count(digitSum)) {
        res = max(res, digitSumOfValue[digitSum] + x);
        digitSumOfValue[digitSum] = max(digitSumOfValue[digitSum], x);
      } else {
        digitSumOfValue[digitSum] = x;
      }
    }

    return res;
  }

private:
  int digitSum(int x) {
    int sum = 0;
    while (x) {
      sum += x % 10;
      x /= 10;
    }

    return sum;
  }
};

File: Daily Challenge/February 2025/Max Sum of a Pair With Equal Sum of Digits.cs
public class Solution {
  public int MaximumSum(int[] nums) {
    Dictionary<int, int> digitSumToMaxValue = new Dictionary<int, int>();
    int result = -1;

    foreach (int x in nums) {
      int digitSum = GetDigitSum(x);

      if (digitSumToMaxValue.ContainsKey(digitSum)) {
        result = Math.Max(result, digitSumToMaxValue[digitSum] + x);
        digitSumToMaxValue[digitSum] = Math.Max(digitSumToMaxValue[digitSum], x);
      } else {
        digitSumToMaxValue[digitSum] = x;
      }
    }

    return result;
  }

  private int GetDigitSum(int x) {
    int sum = 0;

    while (x > 0) {
      sum += x % 10;
      x /= 10;
    }

    return sum;
  }
}